File: Data Structures & Algorithms/find-target-in-rotated-sorted-array/submission-0.py
class Solution:
    def search(self, nums: List[int], target: int) -> int:
        l, r = 0 , len(nums) - 1

        while l <= r:
            mid = (l + r) // 2

            if nums[mid] == target:
                return mid

            if nums[l] <= nums[mid]:
                if target >= nums[l] and target < nums[mid]:
                    r = mid - 1
                else:
                    l = mid + 1
            else:
                # print(l, r)
                if target <= nums[r] and target > nums[mid]:
                    l = mid + 1
                else:
                    r = mid - 1
        
        return -1

File: Data Structures & Algorithms/find-k-closest-elements/submission-0.ts
class Solution {
    /**
     * @param {number[]} arr
     * @param {number} k
     * @param {number} x
     * @return {number[]}
     */

    findClosestElements(arr: number[], k: number, x: number): number[] {
        let left = 0;
        let right = arr.length - k;

        // Binary search to find the optimal starting index of the k-element window
        while (left < right) {
            const mid = Math.floor((left + right) / 2);
            if (x - arr[mid] > arr[mid + k] - x) {
                left = mid + 1;
            } else {
                right = mid;
            }
        }

        return arr.slice(left, left + k);
    }
}
